listenbrainz_spark/stats/user.py

```diff
@@ -0,0 +1,190 @@
+                 , release_name
+                 , release_msid
+                 , release_mbid
+                 , artist_name
+                 , artist_msid
+                 , artist_mbids
+                 , count(release_msid) as cnt
+              FROM %s
+          GROUP BY user_name, release_name, release_msid, release_mbid, artist_name, artist_msid, artist_mbids
+          ORDER BY cnt DESC
+        """ % (table))
+    releases = defaultdict(list)
+    t = time.time()
+    rows = query.collect()
+    releases = defaultdict(list)
+    print("time taken by collect call: %.2f" % (time.time() - t))
+    for row in rows:
+        releases[row.user_name].append({
+            'release_name': row.release_name,
+            'release_msid': row.release_msid,
+            'release_mbid': row.release_mbid,
+            'artist_name': row.artist_name,
+            'artist_msid': row.artist_msid,
+            'artist_mbids': row.artist_mbids,
+            'listen_count': row.cnt,
+        })
+    query_t0 = time.time()
+    print("Query to calculate artist stats processed in %.2f s" % (query_t0 - t0))
+
+
+def get_users(table):
+    """ DataFrame is registered as a temporary table
+    Args:
+        table : name of the temporary table
+
+    Returns:
+        users: list containing user names of all the users
+    """
+    t0 = time.time()
+    query = run_query("""
+            SELECT DISTINCT user_name
+              FROM %s
+        """ % table)
+    users = [row.user_name for row in query.collect()]
+    query_t0 = time.time()
+    query.show()
+    print("Query to get list of users processed in %.2f s" % (query_t0 - t0))
+    return users
+
+def main(app_name):
+    """
+    Args:
+        app_name: Application name to uniquely identify the submitted
+                  application amongst other applications
+    """
+    t0 = time.time()
+    listenbrainz_spark.init_spark_session(app_name)
+    t = datetime.utcnow().replace(day=1)
+    date = t + relativedelta(months=-1)
+    try:
+        df = listenbrainz_spark.sql_context.read.parquet('{}/data/listenbrainz/{}/{}.parquet'.format(config.HDFS_CLUSTER_URI, date.year, date.month))
+        print("Loading dataframe...")
+    except:
+        print ("No Listens for last month")
+        sys.exit(-1)
+    df.printSchema()
+    print(df.columns)
+    print(df.count())
+    table = 'listens_{}'.format(datetime.strftime(date, '%Y_%m'))
+    print(table)
+    df.registerTempTable(table)
+    print("Running Query...")
+    query_t0 = time.time()
+    print("DataFrame loaded in %.2f s" % (query_t0 - t0))
+    users = get_users(table)
+    print("Number of users: %d" % len(users))
+    data = defaultdict(dict)
+    artist_data = get_artists(table)
+    for user_name, artist_stats in artist_data.items():
+        data[user_name]['artists'] = {
+            'artist_stats': artist_stats,
+            'artist_count': len(artist_stats),
+        }
+    recording_data = get_recordings(table)
```

Comment:
Check for exceptions.

Reply:
... here and all the other places in this function ...

Reply:
I am not able to think of any exceptions here. If the parquet does not exist or there is any problem in reading it we would catch it earlier in main, otherwise we would have all the stats. Am I missing out anything?

Reply:
Yes, anytime that something complex such as this is carried it out, it should have a try block around it. If you cannot envisions a problem during such a complex operation, use a generic try block that then prints a stack track and attempts to carry on if it can.